packages/std/src/math/fraction.rs

```diff
@@ -12,3 +12,57 @@ pub trait Fraction<T>: Sized {
     /// If `p` is zero, None is returned.
     fn inv(&self) -> Option<Self>;
 }
+
+impl<T: Clone> Fraction<T> for (T, T) {
+    fn numerator(&self) -> T {
+        self.0.clone()
+    }
+
+    fn denominator(&self) -> T {
+        self.1.clone()
+    }
+
+    fn inv(&self) -> Option<Self> {
+        Some((self.1.clone(), self.0.clone()))
+    }
+}
+
+#[macro_export]
+macro_rules! impl_mul_fraction {
+    ($Uint:ident) => {
+        impl $Uint {
+            pub fn checked_mul_floored<F: Fraction<T>, T: Into<$Uint>>(
```

Comment:
This is a bit more generic than needed, isn't it? It allows things like `Uint256 * Fracttion<u64>`, `Uint256 * Fraction<Uint128>`, ...

How do you feel about restricting it to `F: Fraction<$Uint>` and only support

```
Uint64 * Fraction<Uint64>
Uint128 * Fraction<Uint128>
Uint256 * Fraction<Uint256>
```

I think both ways are correct. One is more generic and one is simpler to document and maintain.

Reply:
At the moment, it allows for potentially a common case (?)

```rust
    // Uint256 * Fraction<Uint128>

    #[test]
    fn mul_ceil_works_with_decimal() {
        let decimal = Decimal::from_ratio(8u128, 21u128);
        let res = Uint256::from(123456u32).mul_ceil(decimal); // 47030.8571
        assert_eq!(Uint256::from(47031u32), res)
    }
```
I'm not sure the maintenance cost, but I figured `T: Into<$Uint>` is kinda self regulating?

I even think it would be nice to support `Uint128* Fraction<Uint256>` by expanding to the largest of the two. Though I think this could be tricker to implement in the macro.

Reply:
Alright, that's fair.

> I even think it would be nice to support Uint128* Fraction<Uint256> by expanding to the largest of the two. Though I think this could be tricker to implement in the macro.

Yeah that would be useful for Uint64*Decimal. I could imagine a using wanting that without being aware that Decimal is a Fraction<128>.

Reply:
To accommodate both directions, I believe the function signature would have to look something like this:

```rust
// assuming a `T * Fraction<U>` would equal a `T`. 
fn mul_floored<F: Fraction<U: Into<T>, T: Into<U>>(rhs: F) -> T;
```

But how would you then be able to do:

```
if T.is_a_larger_bit_size(U) {
   T.full_mul(U.into())
} else {
   U.full_mul(T.into()
}
```

I'm not sure Uint structs are comparative in that way.

Reply:
I think you can use size_of here and compate the two sides. This is `const fn`, i.e. known at compile time, such that the other branch should be optimized away. See https://github.com/CosmWasm/cosmwasm/pull/1574

Reply:
Hmm this is trickier to do in one method than I thought. Here's a trait that would allow the Fraction to expand:
```rust
// Uint256 (self) * Fraction<Uint128> = Uint256
pub trait FractionGrows {
    fn mul_any<F: Fraction<T>, T: Into<Self>>(self, rhs: F) -> Self
    where
        Self: Sized;
}
```
and another that would allow the struct implementing to expand:
```rust
// Uint64 (self) * Fraction<Uint128> = Uint64
pub trait SelfGrows<T>
where
    Self: Into<T>,
{
    fn mul_any<F: Fraction<T>>(self, rhs: F) -> Self;
}
```
If we would be ok with this overloading and the consumer to determine which they needed via the trait import, this would work. However, could we create a single trait bound to accommodate both? In this case:
```rust
// ❌ Both don't implement Into
pub trait BothGrow<T>
where
    Self: Into<T>,
    T: Into<Self>,
{
    fn abc<F: Fraction<T>>(self, rhs: F) -> Self;
}
```
...it requires _both_ to implement the trait. We'd need a trait bound that requires _either_ `Self:Into<T>` OR `T: Into<Self>` to be true. I'm not sure Rust provides the tools for this 🤔 . And then `size_of()` would help us at runtime figure out which one to expand. Maybe casting everything to Uint512 wasn't so bad 😅 .

Reply:
Let's keep it simpel and avoid innecessary complexity. If we say the Fraction's bit size needs to be >= the integer, we are good. A user can any time do `Uint128::from(my_uint64)*the_decimal`.